providers/targeted_refresh.md

```diff
@@ -0,0 +1,180 @@
+      when Vm
+        add_target(:vms, target.ems_ref)
+      end
+    end
+  end
+
+  def add_target(association, ems_ref)
+    return if ems_ref.blank?
+
+    target.add_target(:association => association, :manager_ref => {:ems_ref => ems_ref})
+  end
+
+  # This helps us reference all unique references by collection e.g. all VM targets
+  def references(collection)
+    target.manager_refs_by_association&.dig(collection, :ems_ref)&.to_a&.compact || []
+  end
+end
+```
+
+That will add the `InventoryRefresh::Target` that we can reference in the collector.
+
+Now lets write our `instances` collector method:
+```ruby
+def instances
+  @instances ||= begin
+    references(:vms).map do |ems_ref|
+      compute_client.get_instance(ems_ref)
+    end
+  end
+end
+```
+
+This allows our parser to work exactly the same way, calling `collector.instances` returns an array of instance objects to be parsed.  The parser doesn't even have to know that it is one or two instances versus all of them.
+
+Continue adding targeted methods for the images collection and any others that you wish to be able to target specifically.
+
+It is critical that if your parser depends on other top-level collections (e.g. the instances parser requires that "disks" or "network_ports" be collected from the API) then you _must_ add collector methods for these.  If you don't then targeted refreshes of new inventory will be incomplete and targeted refreshes of existing inventory will delete those dependent collections.
+
+### Testing
+
+Now let's test it out.  First we can do a manual test, we can run this test from a rails console:
+```
+rails c
+ems = ManageIQ::Providers::AwesomeCloud::CloudManager.first
+EmsRefresh.refresh(ems) # Run a full refresh to ensure we are up to date
+vm = ems.vms.first
+EmsRefresh.refresh(vm) # Now we can target individual vms!
+```
+
+You can change something (like the name or description) and confirm that it gets picked up by the targeted refresh.
+
+If you want to target a new vm (e.g. something that isn't in the database yet) then you can use the `InventoryRefresh::Target` notation.  It is a little less convenient to use in quick testing but is better otherwise.
+
+Try creating a new VM in your cloud console and copy its unique reference (just "1234" in this example), then run the following command:
+```
+rails c
+ems = ManageIQ::Providers::AwesomeCloud::CloudManager.first
+target = InventoryRefresh::Target.new(:manager => ems, :association => :vms, :manager_ref => {:ems_ref => "1234"})
+EmsRefresh.refresh(target)
+ems.reload
+ems.vms.find_by(:ems_ref => "1234")
+```
+
+### Conclusion
+
+And that's all you need to get started!  Targeted refresh is most helpful when paired with an event catcher which we will cover next.
+
+Even without an event catcher you can use this (e.g. in a provisioning workflow after the VM) is created in lieu of the full refresh, as this can greatly reduce the time each provision request takes.
```

Comment:
```suggestion
Even without an event catcher you can use this (e.g. in a provisioning workflow after the VM is created) in lieu of the full refresh, as this can greatly reduce the time each provision request takes.
```